pipelines/jenkins/JenkinsAnchoreCICD.txt

```diff
@@ -0,0 +1,98 @@
+            stage("Verifying Anchore Availability") {
+                steps {
+                    script {
+                        // We attempt to discover the availability of the Anchore installation. This is currently looking to see if the api endpoint is reachable. If the call fails
+                        // entirely, we catch the exception and log it, but proceed on. We could add additional checking here on http status or other pieces of the returned json object.
+                        //
+                        // Important point here is anchoreStatus is getting set here so we can guard calls to anchore below and not kill the pipeline if Anchore is unreachable.
+                        try {
+                            withCredentials([usernamePassword(credentialsId: 'Anchore-Pass', passwordVariable: 'pass', usernameVariable: 'user')]) {
+                                env.anchoreStatus = sh(script: 'docker run --name anchorecli --rm anchore/anchore-cli-dev anchore-cli --u $user --p $pass --url $ANCHORE_URL --json system status | jq -r \'.service_states[] | select(.servicename == "apiext") .status_message\'', returnStdout: true).trim()
+                            }
+                            if (anchoreStatus == "available")
+                                println("Anchore API availability status: " + anchoreStatus)
+                        } catch (Exception ex) {
+                                println("ERROR: Error encountered calling the Anchore endpoint.")
+                                anchoreStatus = "unavailable"
+                        }
+                    }
+                }
+            }
+    
+            stage('Submit image for scanning') {
+                steps {
+                    sh 'echo Running anchorecli'
+                    script {
+                        
+                        // Here we are guarding the pipeline...if anchore is not available, we skip this and log a warning.
+                        if (anchoreStatus == "available") {
+                            println("Anchore is availabile...proceeding.")
+                            
+                            // Here we use credentials configured in the Jenkins credentials store to authenticate to the anchore instance.
+                            withCredentials([usernamePassword(credentialsId: 'Anchore-Pass', passwordVariable: 'pass', usernameVariable: 'user')]) {
+                                
+                                // Submit the image to Anchore for scanning
+                                println("Submitting image for analysis...")
+                                env.digest = sh (script: 'docker run --name anchorecli --rm anchore/anchore-cli-dev anchore-cli --u $user --p $pass --url $ANCHORE_URL --json image add $IMAGENAME":"$TAG | jq -r \'.[] .image_detail[] .digest\'',
+                                    returnStdout: true).trim()
+    
+                                // Check on the status of the scan
```

Comment:
I don't think this is strictly necessary, if the image is analyzed already the wait should just return immediately. Is this just for example? I don't think it matters either way so just curious

Reply:
Correct it is not strictly needed @found-it …I’ll add a comment to that effect.